### PR TITLE
Preamp v4 Support

### PR DESCRIPTION
**Pull request by Lohrer**
TODO:
- [x] Verify production tests pass.
- [x] Update changelog with v1.7 changes.
- [x] Finish Preamp EEPROM register documentation.

### CODE REVIEW

File: fw/preamp/CMakeLists.txt [hunk 1]
```diff
@@ -127,7 +115,7 @@ add_custom_command(TARGET ${PROJECT_NAME}.elf POST_BUILD
 
 add_custom_target(program
   COMMAND cd ${CMAKE_CURRENT_SOURCE_DIR}../../.. &&
-          venv/bin/python -m amplipi.hw --flash ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}.bin
+          python -m amplipi.hw --flash ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}.bin
```
Comment (linknum23) [inline fw/preamp/CMakeLists.txt:118]: hmm

File: fw/preamp/preamp.ld [hunk 2]
```diff
@@ -0,0 +1,96 @@
+/* AmpliPi Preamp STM32F030R8 linker script. */
```
Comment (linknum23) [inline fw/preamp/preamp.ld:1]: Why did we make our own?

File: fw/preamp/src/ctrl_i2c.c [hunk 3]
```diff
@@ -297,12 +302,7 @@ void writeReg(uint8_t addr, uint8_t data) {
       }
       break;
 
-    case REG_VOL_ZONE1:
-    case REG_VOL_ZONE2:
-    case REG_VOL_ZONE3:
-    case REG_VOL_ZONE4:
-    case REG_VOL_ZONE5:
-    case REG_VOL_ZONE6: {
+    case REG_VOL_ZONE1 ... REG_VOL_ZONE6: {
```
Comment (linknum23) [inline fw/preamp/src/ctrl_i2c.c:305]: :heart:

**Review by linknum23 — APPROVED** [review]
Rubber Stamp - There are too many changes to review here

**Comment by linknum23** [comment]
Lets do some good real world testing on this

File: amplipi/hw.py [hunk 4]
```diff
@@ -334,20 +333,21 @@ def program(self, filepath: str, unit: int = 0, baud: int = 115200) -> bool:
...
-      # Failed to handshake with the bootloader. Assume unit not present.
-      print(f"Couldn't communicate with {self.unit_num_to_name(unit)}'s bootloader.")
-      plural = 's are' if unit != 1 else ' is'
-      print(f'Assuming only {unit} unit{plural} present and stopping programming')
-
-      # Reset all units to make sure the UART passthrough and
-      # bootloader modes are cleared.
-      self.reset()
-      return False
+    if (unit > 1):
```
[the remark this answers is not in the document]
Comment (Lohrer) [reply amplipi/hw.py]: Silly Python not being C

File: fw/preamp/CMakeLists.txt (hunk 1 above)
Comment (Lohrer) [reply fw/preamp/CMakeLists.txt:118]: What does hmm mean here?

File: fw/preamp/src/audio.c [hunk 5]
```diff
@@ -183,12 +191,12 @@ void setZoneSource(size_t zone, size_t src) {
 
   // Disconnect zone from all sources first
   for (size_t src = 0; src < NUM_SRCS; src++) {
-    writePin(zone_src_[zone][src], false);
+    pin_write(zone_src_[zone][src], !mux_en_level_);
```
[the remark this answers is not in the document]
Comment (Lohrer) [reply fw/preamp/src/audio.c:194]: Hmm, I like where that's going but I think that's going to have to be something to keep in mind for the next update. I looked at making that change and would have to touch both I2Cs, audio, and the initialization.

File: fw/preamp/CMakeLists.txt (hunk 1 above)
Comment (rtertiaer) [reply fw/preamp/CMakeLists.txt:118]: an option here that might work well is calling python like so:
```
/usr/bin/env python3
```
this ensures it's python3 that is called, and also tolerates being called from within a virtualenv so you don't have to hardcode any paths.

**Pull request merged**